2277 Add items to solr for recap.email and extract missed recap documents

albertisfu opened the pull request:
I've added the `add_items_to_solr` task after extracting recap.email documents.

Also added a new option `extract_missed_rd` to `cl_update_index` command:

`manage.py cl_update_index --extract_missed_rd`

This option will query recap documents that need content extraction and extract them.

After analyzing the existing code, I think the best option to retrieve recap documents that need content extraction is to use a query that takes the conditions of the `needs_extraction` property and also evaluating the document `text_plain` content passing it to `needs_ocr`. In the beginning, I was just thinking to do something like  `text_plain=""` to only extract documents that already don't have content but I think is better to use `needs_ocr` so we'll also consider documents that need OCR, this because seems that there was a bug assigning `ocr_status` after extraction here: https://github.com/freelawproject/courtlistener/blob/main/cl/scrapers/tasks.py#L211

I think that line:

```
case True, False:
       if not ocr_available:
             rd.ocr_status = RECAPDocument.OCR_UNNECESSARY
```
Might be:

```
ocr_needed = needs_ocr(content)
...
case True, False:
       if not ocr_needed:
              rd.ocr_status = RECAPDocument.OCR_UNNECESSARY
                        
```

Because using `ocr_available` that by default is `True` and also it doesn't indicate if `OCR_UNNECESSARY` should be assigned. So in many cases the `ocr_status` remains as `None` even though the content is extracted and doesn't need OCR.

So I think the right condition is to use: `needs_ocr`

I also found another possible bug with `needs_ocr` since it looks for good content besides the header stamp:
`Case 2:06-cv-00376-SRW Document 1-2 Filed 04/25/2006 Page 1 of 1`

Seems that it won't work for appellate documents, so I checked the bunch of appellate documents I have and found another possible types of headers like:

```
Appeal: 15-1504 Doc: 6 Filed: 05/12/2015 Pg: 1 of 4
Appellate Case: 14-3253 Page: 1 Date Filed: 01/14/2015 Entry ID: 4234486
USCA Case #16-1062 Document #1600692 Filed: 02/24/2016 Page 1 of 3
USCA11 Case: 21-12355 Date Filed: 07/13/202 Page: 1 of 2
```
So in order to also consider these cases I added a tuple of starting words: 
`("Case", "Appellate", "Appeal", "USCA", "USCA11")`

I saw that `cl_update_index` already has the `add_or_update_all` option to add missed objects to Solr. So I didn't add an additional method for it. Or should I add to `extract_missed_rd` as a second step a method to add to Solr  documents that just were extracted?

mlissner commented on cl/search/management/commands/cl_update_index.py:
```diff
@@ -168,11 +214,23 @@ def add_arguments(self, parser):
         )
 
+        parser.add_argument(
+            "--extract_missed_rd",
```
```suggestion
            "--extract-unextracted-rds",
```

albertisfu replied:
changed by: `extract-and-add-solr-unextracted-rds`

mlissner commented on cl/lib/recap_utils.py:
```diff
@@ -101,7 +105,7 @@ def needs_ocr(content):
     for line in content.splitlines():
         line = line.strip()
-        if line.startswith("Case"):
+        if line.startswith(("Case", "Appellate", "Appeal", "USCA", "USCA11")):
```
```suggestion
        if line.startswith(("Case", "Appellate", "Appeal", "USCA")):
```

mlissner reviewed (commented):
Cool. A few thoughts:

1. The extraction stuff does look a little out of place to me in the `cl_update_index` command. I don't care enough to change it, but if it bugs you too, maybe go for it.
2. Thanks for the tweaks to the needs_ocr function. Nice.
3. For the tweak to the case statement in the extract_recap_pdf method, yes, that looks right. Would it make your cl_update_index query easier if we just updated the DB? For example, anything that has `ocr_needed` of `null` and that has plaintext could be set to `OCR_UNNCESSARY`? Rather than working around bad data, can we fix it?
4. `add_or_update_all` is an absolute *monster* of a function. It will update millions of items. We should find a more surgical way to just update the ones we care about.

albertisfu commented:
Great!

A couple of questions:

> The extraction stuff does look a little out of place to me in the cl_update_index command. I don't care enough to change it, but if it bugs you too, maybe go for it.

Yeah, I think it might be in a better place, do you have an idea in mind on which command might it fit better?
I saw that there is this one `reprocess_recap_dockets` or might be better to create a new one?

> For the tweak to the case statement in the extract_recap_pdf method, yes, that looks right. Would it make your cl_update_index query easier if we just updated the DB? For example, anything that has ocr_needed of null and that has plaintext could be set to OCR_UNNCESSARY? Rather than working around bad data, can we fix it?

Yeah, if we fix it that would simplify the query so that we won't need to check again if there is there's a good `plain_text.`
So for this fix, should we add this code also into a command, or just writing a code snippet to apply via the django console would be enough? 


Thanks for your comments.

mlissner commented:
> I saw that there is this one reprocess_recap_dockets

Sure, that seems like a better place for it.

>  just writing a code snippet to apply via the django console would be enough?

A code snippet is fine. Much easier!

albertisfu commented:
Thank you, I've moved the command to `reprocess_recap_dockets`

`manage.py reprocess_recap_dockets --extract-and-add-solr-unextracted-rds`

Also added the `add_items_to_solr` task to this command, so all the recap documents being extracted are going to be also added
to solr.

This is the code snippet to fix all the recap documents that failed to assign the `OCR_UNNECESSARY` status.

```
from cl.search.models import RECAPDocument
from cl.lib.recap_utils import needs_ocr
from django.db.models import Q


rds_to_fix_ocr_status = [
        x
        for x in RECAPDocument.objects.filter(
            (Q(ocr_status=None) | Q(ocr_status=RECAPDocument.OCR_NEEDED))
            & Q(is_available=True)
            & ~Q(filepath_local="")
        )
        if not needs_ocr(x.plain_text)
    ]

for rd in rds_to_fix_ocr_status:
		rd.ocr_status = RECAPDocument.OCR_UNNECESSARY
		rd.save()
        
print(f"Set as OCR_UNNECESSARY: {len(rds_to_fix_ocr_status)} documents")
```

So within the `extract-and-add-solr-unextracted-rds` query now we avoid checking if the document has `plain_text` and if `needs_ocr`, so the query is simpler only reproducing the conditions of `needs_extraction` property.

Related to this, testing this command and code snippet I downloaded a couple of PDF that need OCR from CL and did some tests.

Something weird happened with doctor, the couple of documents that I tested were not extracted by OCR and I got a `406` response. 

And if you look at this API request: https://www.courtlistener.com/api/rest/v3/recap-documents/?ocr_status=1&ordering=date_created

Seems that we haven't extracted OCR documents since April, is that normal?

I think that's the reason there are some documents with a plain_text like this:
```
Case 1:16-mj-00487-CFH Document 1 Filed 09/26/16 Page 1 of 25
Case 1:16-mj-00487-CFH Document 1 Filed 09/26/16 Page 2 of 25
Case 1:16-mj-00487-CFH Document 1 Filed 09/26/16 Page 3 of 25
Case 1:16-mj-00487-CFH Document 1 Filed 09/26/16 Page 4 of 25
Case 1:16-mj-00487-CFH Document 1 Filed 09/26/16 Page 5 of 25
```

mlissner commented on cl/recap/management/commands/reprocess_recap_dockets.py:
```diff
@@ -1,11 +1,59 @@
+            throttle.maybe_wait()
+            chain(
+                extract_recap_pdf.si(chunk).set(queue=queue),
+                add_items_to_solr.s("search.RECAPDocument"),
```
Don't forget the queue here. :)

albertisfu replied:
yeah, added.

mlissner reviewed (changes requested):
One small tweak, please. 

It's not great that we haven't done OCR since April. Amazing that we haven't caught that before now. Do you think your fixes here will fix that, or do you think it needs something else as well?

albertisfu commented:
Thanks, I've added the queue.

About OCR, no the OCR problem seems that is not related to `extract_recap_pdf` task. Based on some tests I think the error might be related to the microservice call or a doctor issue. 

It seems a different issue to check.

mlissner commented:
Cool. Will you file a fresh issue about the OCR problem, please? I'll wait for this to deploy, then run your code snippet.

mlissner commented:
Hm, well, I think the code ran out of memory, which is interesting because that pod has 1G, but I rewrote it to iterate instead:

```
In [8]: for i, rd in enumerate(RECAPDocument.objects.filter(
   ...:    ...:             (Q(ocr_status=None) | Q(ocr_status=RECAPDocument.OCR_NEEDED))
   ...:    ...:             & Q(is_available=True)
   ...:    ...:             & ~Q(filepath_local="")
   ...:    ...:         ).iterator()):
   ...:     if not needs_ocr(rd.plain_text):
   ...:         rd.ocr_status = RECAPDocument.OCR_UNNECESSARY
   ...:         rd.save()
   ...:     print(f"Did item {i}: {rd}")
```

Hopefully that'll do it. There's about 300k docs it's going to run against.

mlissner commented:
I'm running this script currently and expect it to finish soon. Poor old doctor is running 32 pods, all of them cranking along. I could scale up, but I'm a bit lazy.

mlissner commented:
OK, getting this error a lot: 

```
 cl-celery-prefork-bulk-5b6448fc7d-8c8b4 Traceback (most recent call last):                                                                                                                                                                                  │
│ cl-celery-prefork-bulk-5b6448fc7d-8c8b4   File "/usr/local/lib/python3.10/site-packages/celery/app/trace.py", line 412, in trace_task                                                                                                                       │
│ cl-celery-prefork-bulk-5b6448fc7d-8c8b4     R = retval = fun(*args, **kwargs)                                                                                                                                                                               │
│ cl-celery-prefork-bulk-5b6448fc7d-8c8b4   File "/usr/local/lib/python3.10/site-packages/sentry_sdk/integrations/celery.py", line 204, in _inner                                                                                                             │
│ cl-celery-prefork-bulk-5b6448fc7d-8c8b4     reraise(*exc_info)                                                                                                                                                                                              │
│ cl-celery-prefork-bulk-5b6448fc7d-8c8b4   File "/usr/local/lib/python3.10/site-packages/sentry_sdk/_compat.py", line 54, in reraise                                                                                                                         │
│ cl-celery-prefork-bulk-5b6448fc7d-8c8b4     raise value                                                                                                                                                                                                     │
│ cl-celery-prefork-bulk-5b6448fc7d-8c8b4   File "/usr/local/lib/python3.10/site-packages/sentry_sdk/integrations/celery.py", line 199, in _inner                                                                                                             │
│ cl-celery-prefork-bulk-5b6448fc7d-8c8b4     return f(*args, **kwargs)                                                                                                                                                                                       │
│ cl-celery-prefork-bulk-5b6448fc7d-8c8b4   File "/usr/local/lib/python3.10/site-packages/celery/app/trace.py", line 704, in __protected_call__                                                                                                               │
│ cl-celery-prefork-bulk-5b6448fc7d-8c8b4     return self.run(*args, **kwargs)                                                                                                                                                                                │
│ cl-celery-prefork-bulk-5b6448fc7d-8c8b4   File "/usr/local/lib/python3.10/site-packages/celery/app/autoretry.py", line 35, in run                                                                                                                           │
│ cl-celery-prefork-bulk-5b6448fc7d-8c8b4     return task._orig_run(*args, **kwargs)                                                                                                                                                                          │
│ cl-celery-prefork-bulk-5b6448fc7d-8c8b4   File "/opt/courtlistener/cl/scrapers/tasks.py", line 188, in extract_recap_pdf                                                                                                                                    │
│ cl-celery-prefork-bulk-5b6448fc7d-8c8b4     response = microservice(                                                                                                                                                                                        │
│ cl-celery-prefork-bulk-5b6448fc7d-8c8b4   File "/opt/courtlistener/cl/lib/microservice_utils.py", line 54, in microservice                                                                                                                                  │
│ cl-celery-prefork-bulk-5b6448fc7d-8c8b4     "file": (item.filepath_local.name, item.filepath_local.read())                                                                                                                                                  │
│ cl-celery-prefork-bulk-5b6448fc7d-8c8b4   File "/usr/local/lib/python3.10/site-packages/django/core/files/utils.py", line 42, in <lambda>                                                                                                                   │
│ cl-celery-prefork-bulk-5b6448fc7d-8c8b4     read = property(lambda self: self.file.read)                                                                                                                                                                    │
│ cl-celery-prefork-bulk-5b6448fc7d-8c8b4   File "/usr/local/lib/python3.10/site-packages/django/db/models/fields/files.py", line 45, in _get_file                                                                                                            │
│ cl-celery-prefork-bulk-5b6448fc7d-8c8b4     self._file = self.storage.open(self.name, 'rb')                                                                                                                                                                 │
│ cl-celery-prefork-bulk-5b6448fc7d-8c8b4   File "/usr/local/lib/python3.10/site-packages/django/core/files/storage.py", line 38, in open                                                                                                                     │
│ cl-celery-prefork-bulk-5b6448fc7d-8c8b4     return self._open(name, mode)                                                                                                                                                                                   │
│ cl-celery-prefork-bulk-5b6448fc7d-8c8b4   File "/usr/local/lib/python3.10/site-packages/storages/backends/s3boto3.py", line 430, in _open                                                                                                                   │
│ cl-celery-prefork-bulk-5b6448fc7d-8c8b4     raise FileNotFoundError('File does not exist: %s' % name)                                                                                                                                                       │
│ cl-celery-prefork-bulk-5b6448fc7d-8c8b4 FileNotFoundError: File does not exist: recap/gov.uscourts.txed.195194/gov.uscourts.txed.195194.5.11.pdf  
```
I'm going to open a new issue for it. Looks like a lot of these are wrong about the doc being available and we should fix that if so.
